rootfs/etc/nginx/template/nginx.tmpl

```diff
@@ -663,15 +663,15 @@ stream {
         {{ $all := .First }}
         {{ $server := .Second }}
         {{ range $address := $all.Cfg.BindAddressIpv4 }}
-        listen {{ $address }}:{{ $all.ListenPorts.HTTP }}{{ if $all.Cfg.UseProxyProtocol }} proxy_protocol{{ end }}{{ if eq $server.Hostname "_"}} default_server {{ if $all.Cfg.ReusePort }}reuseport{{ end }} backlog={{ $all.BacklogSize }}{{end}};
+        listen {{ $address }}:{{ $all.ListenPorts.HTTP }}{{ if $all.Cfg.UseProxyProtocol }} proxy_protocol{{ end }}{{ if eq $server.Hostname "_"}} default_server {{ if $all.Cfg.ReusePort }}reuseport{{ end }} backlog={{ $all.BacklogSize }}{{end}} {{ if $all.Cfg.UseHTTP2 }}http2{{ end }};
```

Comment:
@aledbf Hi. Thanks as always for maintaining the project!

I am still sticking to support this in ingress-nginx, after https://github.com/kubernetes/ingress-nginx/issues/2444#issuecomment-387437204, and seeing that the author of nginx grpc module is pretty negative to fix this in nginx.

You said:

> We can add this feature only if:
>
> * there is only one path exposed in the hostname
> * all the paths use gRPC

Then, would it be acceptable if this `http2` option is appended either:

- there are two or more `ingress.Location`, but all the `ingress.Location.GRPC` are `true` for the host

Or:

- there is just one `ingress.Location` for the host

?

Reply:
@aledbf Thanks. Ah, so the number of `ingress.Location`s doesn't matter at all?
We should just decide whether http2 should be enabled or not here, by checking every `ingress.Location.GRPC` is true, right?

Reply:
right

Reply:
Update; I have tested this by running grpcurl against nginx ingress whose backend is grpc server built with the go-grpc from https://github.com/grpc/grpc-go/pull/2045

Surprisingly, it turned out that I don't need `listen http2` at all to make it actually work.

My working nginx.conf looks like this:

```
# cat nginx.conf | grep listen
		listen 80 default_server  backlog=511 http2;
		listen [::]:80 default_server  backlog=511 http2;
		listen 443  default_server  backlog=511 ssl http2;
		listen [::]:443  default_server  backlog=511 ssl http2;
		listen 80;
		listen [::]:80;
		listen 80;
		listen [::]:80;
		listen 18080 default_server  backlog=511;
		listen [::]:18080 default_server  backlog=511;
```

Reply:
@cvanderschuere Hi. Thanks for the fix! But for me this change wasn't necessarily to make my plantext grpc(a.k.a grpc over h2c) work.

Did it actually work for you? How did you test it?

Reply:
@aledbf Sorry for bothering but would you mind confirming this? I'm confused

Reply:
Sorry, it turned out I was misreading my nginx.conf. This change is definitely needed in order to make h2c work.
I'll go ahead with implementing https://github.com/kubernetes/ingress-nginx/pull/2465/files#r187064310